gossip: deflake TestClientGossipMetrics

d4l3k opened the pull request:
Closes https://github.com/cockroachdb/cockroach/issues/8525

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8607)

<!-- Reviewable:end -->

d4l3k committed:
gossip: deflake TestClientGossipMetrics

tamird commented on gossip/client_test.go:
```diff
@@ -216,17 +216,22 @@ func TestClientGossipMetrics(t *testing.T) {
+			// Since there are two gossip nodes, there should be either one incoming
+			// or outgoing connection.
+			for i, g := range []*Gossip{local, remote} {
+				s := g.server
```
No need for this alias, `server` is embedded.

tamird commented:
What is the root cause here? Why isn't the connection always bidirectional?

tamird commented:
Ah, it's because gossip de-duplicates connections.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[gossip/client_test.go, line 219 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-KPOWajVf8bbQ8yafHsb:-KPOWajVf8bbQ8yafHsc:b-2ketgb) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfcd8433d4d50c2fe51293d8e43a471bc589d5d8/gossip/client_test.go#L219)):*
> ```Go
> 			}
> 
> 			// Since there are two gossip nodes, there should be either one incoming
> ```

There should be exactly one incoming or outgoing connection.

This comment should mention that this is due to gossip's connection de-duplication.

---

*[gossip/client_test.go, line 233 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-KPOWmaG_wKQYP0AvZsm:-KPOWmaG_wKQYP0AvZsn:bmrpycd) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfcd8433d4d50c2fe51293d8e43a471bc589d5d8/gossip/client_test.go#L233)):*
> ```Go
> 					count += gauge.Value()
> 				}
> 				if count <= 0 {
> ```

this should be checking that count is exactly equal to 1.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-:-KPOWXXAWTqWslSMGM_1:br9t6pr)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Previously this test worked due to the initial instability of the gossip network. Fixed now.

I also cleaned up the rest of the test. There wasn't any need to add the infos multiple times or two clients. That was just due to the fact the gossip server metrics wasn't passed into the client.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[gossip/client_test.go, line 219 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-KPOWajVf8bbQ8yafHsb:-KPOYNnpfqFlhG8lh8YH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfcd8433d4d50c2fe51293d8e43a471bc589d5d8/gossip/client_test.go#L219)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> There should be exactly one incoming or outgoing connection.
> 
> This comment should mention that this is due to gossip's connection de-duplication.
</details>
Done.

---

*[gossip/client_test.go, line 222 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-KPORWi0JnTMHX94WCgC-r1-222:-KPOYOL0UiaDhV64xWsN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfcd8433d4d50c2fe51293d8e43a471bc589d5d8/gossip/client_test.go#L222)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> No need for this alias, `server` is embedded.
</details>
Done.

---

*[gossip/client_test.go, line 233 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-KPOWmaG_wKQYP0AvZsm:-KPOYOulBfcMmxU4lV-z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfcd8433d4d50c2fe51293d8e43a471bc589d5d8/gossip/client_test.go#L233)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this should be checking that count is exactly equal to 1.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-:-KPOYPcEJJr3-D1t01jp:b9gwsax)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[gossip/client_test.go, line 219 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-KPOWajVf8bbQ8yafHsb:-KPO__XELCnlksCbcuV9:b-9059lw) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfcd8433d4d50c2fe51293d8e43a471bc589d5d8/gossip/client_test.go#L219)):*
> This comment should mention that this is due to gossip's connection de-duplication.

---

*[gossip/client_test.go, line 233 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-KPOWmaG_wKQYP0AvZsm:-KPO_h4Ho_LJ9tOckQlF:blfkcko) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfcd8433d4d50c2fe51293d8e43a471bc589d5d8/gossip/client_test.go#L233)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Done.
</details>
```
if expected := 1; count != expected {
```

then you can `%d` `expected` into the error below

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[gossip/client_test.go, line 219 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-KPOWajVf8bbQ8yafHsb:-KPOba_nQT6d1QvHZyjc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfcd8433d4d50c2fe51293d8e43a471bc589d5d8/gossip/client_test.go#L219)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> > This comment should mention that this is due to gossip's connection de-duplication.
</details>
Done.

---

*[gossip/client_test.go, line 233 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-KPOWmaG_wKQYP0AvZsm:-KPOahRByJw-fjgMWscn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfcd8433d4d50c2fe51293d8e43a471bc589d5d8/gossip/client_test.go#L233)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ```
> if expected := 1; count != expected {
> ```
> 
> then you can `%d` `expected` into the error below
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[gossip/client_test.go, line 222 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-KPORWi0JnTMHX94WCgC-r1-222:-KPOfzh9P3G_t1nOo8Pc:b2afd8b) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfcd8433d4d50c2fe51293d8e43a471bc589d5d8/gossip/client_test.go#L222)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Done.
</details>
Is holding these mutexes until the end of the function really the intention here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[gossip/client_test.go, line 233 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-KPOgYBFkDKewmtlGyfl:-KPOgYBFkDKewmtlGyfm:b-xawst1) ([raw file](https://github.com/cockroachdb/cockroach/blob/eadc6d9481f4e506c6dbc341285f2cb449e6b2a9/gossip/client_test.go#L233)):*
> ```Go
> 				const expected = 1
> 				if count != expected {
> 					return errors.Errorf("%d: expected metrics incomig + outgoing connection count == %d; = %d", i, expected, count)
> ```

incoming

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[gossip/client_test.go, line 222 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-KPORWi0JnTMHX94WCgC-r1-222:-KPOm5H-jkGFXvuqDnHd:bq15x7h) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfcd8433d4d50c2fe51293d8e43a471bc589d5d8/gossip/client_test.go#L222)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Is holding these mutexes until the end of the function really the intention here?
</details>
That lock is just held in this retry loop function. It's not a problem as no processing needs to happen in the loop.

---

*[gossip/client_test.go, line 233 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-KPOgYBFkDKewmtlGyfl:-KPOo4rHymUsWZg_fdA0:bbvswh8) ([raw file](https://github.com/cockroachdb/cockroach/blob/eadc6d9481f4e506c6dbc341285f2cb449e6b2a9/gossip/client_test.go#L233)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> incoming
</details>
My bad fixed in https://github.com/cockroachdb/cockroach/pull/8610

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[gossip/client_test.go, line 222 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-KPORWi0JnTMHX94WCgC-r1-222:-KPOpRxJKjypLPVFkBf7:bkg022m) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfcd8433d4d50c2fe51293d8e43a471bc589d5d8/gossip/client_test.go#L222)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> That lock is just held in this retry loop function. It's not a problem as no processing needs to happen in the loop.
</details>
It can be a problem if some other code might be grabbing the same mutexes in a different order. I know it's not the case here but it's a bad pattern IMO

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
*[gossip/client_test.go, line 222 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-KPORWi0JnTMHX94WCgC-r1-222:-KPOsyWT85bilihIXfpC:blkt3vo) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfcd8433d4d50c2fe51293d8e43a471bc589d5d8/gossip/client_test.go#L222)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> It can be a problem if some other code might be grabbing the same mutexes in a different order. I know it's not the case here but it's a bad pattern IMO
</details>
Fair enough. There wouldn't be a case of out of order locking because there's only one lock in this case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[gossip/client_test.go, line 222 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-KPORWi0JnTMHX94WCgC-r1-222:-KPOuYT4SW4K5VO2D_B9:b-2zluj9) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfcd8433d4d50c2fe51293d8e43a471bc589d5d8/gossip/client_test.go#L222)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Fair enough. There wouldn't be a case of out of order locking because there's only one lock in this case.
</details>
What do you mean there is only one lock? `local.server.mu` and `global.server.mu` no?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[gossip/client_test.go, line 222 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-KPORWi0JnTMHX94WCgC-r1-222:-KPOx2631NEz_eDANhNI:b-1xtpnd) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfcd8433d4d50c2fe51293d8e43a471bc589d5d8/gossip/client_test.go#L222)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> What do you mean there is only one lock? `local.server.mu` and `global.server.mu` no?
</details>
Meh, I meant  `local.server.mu` and `remote.server.mu`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
*[gossip/client_test.go, line 222 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607#-KPORWi0JnTMHX94WCgC-r1-222:-KPPLfdnzZttlUaXr7Pn:bi0c7ns) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfcd8433d4d50c2fe51293d8e43a471bc589d5d8/gossip/client_test.go#L222)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Meh, I meant  `local.server.mu` and `remote.server.mu`
</details>
There is only be one gossip.server object per cockroach node. Thus, nothing should be able to lock both of them.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8607)*
<!-- Sent from Reviewable.io -->
